### PR TITLE
Separate hotcounters

### PR DESCRIPTION
**Pull request by fsfod**
Switch from function and loop hot counters being in shared small(64 slot) hash table with the hashslot being determined by the address of the bytecode, to being in the function prototype for function hot counters and for the loop counters insert a dummy bytecode after the loop bytecode to contain the counter.

### CODE REVIEW

**Comment by ltratt** [comment]
Can you squash these commits down to a sensible subset before review please? Or is it still WIP given the last commit?

**Comment by fsfod** [comment]
The last commit is temporary system to help verify stuff is getting hot at the right call/loop hot values and to later on compare to normal LuaJIT. I can squash some stuff down but i don't want to lose to much context on what some of the changes are doing.

**Comment by ltratt** [comment]
If you can squash away things which are definitely not helpful, that would be good.

**Comment by fsfod** [comment]
@ltratt i squashed most of the fixes in to the main commits

**Comment by fsfod** [comment]
@ltratt The TEMP commits are only for testing and will be removed for the final code

File: src/lj_trace.c [hunk 1]
```diff
@@ -736,7 +749,14 @@ void LJ_FASTCALL lj_trace_hot(jit_State *J, const BCIns *pc)
   /* Note: pc is the interpreter bytecode PC here. It's offset by 1. */
   ERRNO_SAVE
   /* Reset hotcount. */
-  hotcount_set(J2GG(J), pc, J->param[JIT_P_hotloop]*HOTCOUNT_LOOP);
+  if (bc_op(pc[-1]) >= BC_FUNCF && bc_op(pc[-1]) <= BC_JFUNCV) {
+    GCproto *pt = (GCproto *)(((char *)(pc-1)) - sizeof(GCproto));
+    lua_assert(pt->hotcount == 0xffff);
```
Comment (ltratt) [inline src/lj_trace.c:760]: How is this assert true? I don't see where `0xffff` is written anywhere.

**Comment by ltratt** [comment]
I had a quick look at the first commit: the commit messages need a lot more detail in (e.g. what's the data layout? what extra things have been added? what are the trade-offs) before I can give useful details. If you can do that, I'll try and turn a review round quickly.

File: src/lj_trace.c (hunk 1 above)
Comment (fsfod) [reply src/lj_trace.c:760]: The hot counter counts down to 0 when it underflows [here in vm_x86.dasc](https://github.com/softdevteam/LuaJIT/pull/11/files#diff-f3a7bbc53d3494d691a380e7bafdd7a5R448) it calls lj_dispatch_call that calls this function.

Comment (ltratt) [reply src/lj_trace.c:760]: But where is it initialised?

Comment (fsfod) [reply src/lj_trace.c:760]: It initialised when a function is parsed and its function prototype created in [fs_finish](https://github.com/softdevteam/LuaJIT/pull/11/files#diff-f49c5e8d3541b178bcf63efa1a74216aR1581)

Comment (ltratt) [reply src/lj_trace.c:760]: Does that mean the assert should be `lua_assert(pt-hotcount == L2J(ls->L)->param[JIT_P_hotfunc])`?

Comment (fsfod) [reply src/lj_trace.c:760]: No it will always be a 16 bit -1 coming from the interpreter in to this function

Comment (ltratt) [reply src/lj_trace.c:760]: OK.

**Comment by fsfod** [comment]
@ltratt The code should be ready to merge the only new things are the last commit that got some fixes and the redesigned unit tests for the counters in test/counters.lua that no longer read the raw counters from a temporary debug API that i've now removed.

**Comment by ltratt** [comment]
Can you squash these down before review and also say whether we need to run new performance numbers or not given these changes?

**Comment by fsfod** [comment]
The commits are mostly squashed down to what i wanted to keep preserve the context of the changes.
I don't think we really need to run new performance numbers, the defaults for the new JIT parameters should recreate the same behaviour of the original code which is 11 attempts for a function or loop to start a root trace before being blacklisted with the same counter backoff scaling.

**Comment by ltratt** [comment]
Mostly squashed down?

Why 11 for the starting count for functions/loops? Did you try any alternative numbers?

**Comment by fsfod** [comment]
I've probably said this before but LuaJIT's 16 bit hot counters count down to zero and trigger a root trace when they underflow zero. 11 was not the starting count its the number of times a hotcounter can be reset to a higher and higher starting count after hitting zero and failing to make a valid trace.

**Comment by ltratt** [comment]
Point taken: I misread this was about blacklisting. But that doesn't invalidate my original question: have we tried different values for the compilation threshold?

Also, is this PR fully squashed for review or mostly squashed?

**Comment by fsfod** [comment]
No we haven't, that's part of why i separated the JIT parameters for loops and functions was so we could try out different values for functions and loops instead functions always taking twice as long to get hot as loops. I do have some worry the results could be highly benchmark specific from the order that traces are generated changing.

I consider it fully squashed for review.

**Comment by ltratt** [comment]
Before we think about merging this, I think we should have performance numbers (maybe from slightly shortened runs) with different values. And, yes, they'll definitely be specific to our benchmark suite -- but at least our benchmark suite isn't completely tiny. Or, looked at another way, there isn't a better benchmark suite that we know of!

**Comment by fsfod** [comment]
@ltratt So the separate counters starting hot count was effectively off by one vs the original shared counter system. These two new commits should fix that.

**Comment by ltratt** [comment]
My only (very) minor comment is: can we consistent with the spacing around `-`? We currently have `f-1` and `'f - 1`. Let's stick with one (preferably the latter).

**Comment by fsfod** [comment]
@ltratt I've tried to make it more consistent

**Comment by ltratt** [comment]
I fear it's got worse :( `rand_total*2 + random_backoff`, for example, is just horrible. Let's consistently do `a + b`.

**Comment by fsfod** [comment]
Most LuaJIT's Lua code is written without any spacing for operators https://github.com/LuaJIT/LuaJIT-test-cleanup/blob/master/test/trace/snap.lua

**Comment by ltratt** [comment]
Even that file has `a % b`, so it's not consistent. Let's do better.

**Comment by fsfod** [comment]
@ltratt Are you happy with it now?

**Comment by ltratt** [comment]
Yes, but we still need the benchmarking numbers before we decide to merge.

**Comment by fsfod** [comment]
@ltratt squashed and ready to merge.

**Review by ltratt — APPROVED** [review]

**Pull request merged**